Feature/implementation tester (#345)
multiscale/muscle3: docs/source/muscle_tester.rst

.. _muscle-tester:

==========================
Testing your MUSCLE3 model
==========================

When developing a MUSCLE3 component, it is useful to be able to test it in
isolation — without having to set up and run the full coupled simulation. The
:class:`~libmuscle.pytest.MuscleTester` class (together with the

> ``muscle3_tester`` pytest fixture) makes this easy: it starts a real MUSCLE3

Reviewer: Explain pytest and fixture?

Reviewer: Note that users don't necessarily have `pytest` installed after installing muscle3

manager in the background and connects a *tester component* to all ports of
your implementation, so you can send and receive messages programmatically from
within a pytest test.

How it works
============

Given a yMMSL file that describes your implementation (its ports, executable,
etc.), :class:`~libmuscle.pytest.MuscleTester`:

1. Reads the yMMSL configuration and automatically adds a *tester component*
   that is wired to every port of the implementation under test.
2. Starts a real MUSCLE3 manager in a background process.
3. Returns an :class:`~libmuscle.pytest.implementation_tester.ImplementationTester`
   object that you can use to :meth:`~libmuscle.pytest.implementation_tester.ImplementationTester.send`
   and :meth:`~libmuscle.pytest.implementation_tester.ImplementationTester.receive`

>    messages on those ports, just as the real coupled component would.

Reviewer: I'm not sure if this belongs in the user documentation.

If you want to include it, probably better to do it all the way at the end. The main goal for this doc page should be: how to use this feature.

Quick start
===========

Step 1: Provide the yMMSL file for testing your implementation
---------------------------------------------------

Your implementation needs a yMMSL file that declares its ports. For example,
for a simple micro model that receives a value on ``init`` and sends a result
on ``final``:

.. code-block:: yaml
    :caption: micro.ymmsl

    ymmsl_version: v0.2

    programs:
      micro:
        ports:
          f_init: init
          o_f: final
        executable: python3

>         args: "-m micro"

Reviewer: Let's provide an example pointing to a python file instead of a python module?

    resources:
      micro:
        threads: 1

Step 2: Use the ``muscle3_tester`` fixture in your test
-------------------------------------------------------

.. code-block:: python
    :caption: test_micro.py

    from libmuscle import Message
    from libmuscle.pytest import MuscleTester

    def test_micro_model(muscle3_tester: MuscleTester) -> None:
        """Test the micro model by acting as the macro."""

>         tester = muscle3_tester.start_implementation("micro.ymmsl", "micro")

Reviewer: This makes some assumptions on the file/directory layout. Can you include that somewhere as well? I'd imagine these files to live in a separate `tests` folder and you'd run `pytest` from the project root folder?

Reviewer 2: Is there a use case for inlining the ymmsl? I haven't really thought about how this would be used, but that would solve the need-to-find-that-file problem.

Author: `start_implementation` currently only accepts a file path string. To support inlining, the method would need to be extended to also accept a `ymmsl.v0_2.Configuration` object directly, right?

Reviewer: > start_implementation currently only accepts a file path string. To support inlining, the method would need to be extended to also accept a ymmsl.v0_2.Configuration object directly, right?

It might be nice to support similar arguments as `ymmsl.load_as`:
- If argument is a `str`, it should be a string containing the ymmsl (so the yMMSL can be inline in the test file, see example below)
- If argument is a pathlib.Path, it should point to a file containing the ymmsl

Example:
```python
config = """
ymmsl_version: v0.2
description: yMMSL config for testing xyz implementation
implementations:
  xyz:
    executable: /path/to/xyz
"""

def test_xyz(muscle3_tester):
    tester = muscle3_tester.start_implementation(config, "xyz")
    ...
```

        # Send a message to the micro model's 'init' port
        tester.send("init", Message(0.0, 10.0, 42))

        # Receive the result from the micro model's 'final' port
        reply = tester.receive("final")

        assert reply.data == 42

Step 3: Run your tests
----------------------

.. code-block:: bash

    pytest test_micro.py

Timeouts and error handling
============================

By default, :meth:`~libmuscle.pytest.MuscleTester.start_implementation`
uses a 60-second timeout for all receive operations. If the implementation
does not send a message within that time, a :class:`RuntimeError` is raised

> and the test fails. You can adjust this timeout:

Reviewer: Explain why you'd want to do this? E.g. long-running calculations?

.. code-block:: python

    tester = muscle3_tester.start_implementation(
        "micro.ymmsl", "micro", default_timeout=5.0
    )

You can also override the timeout for individual receive calls:

.. code-block:: python

    reply = tester.receive("final", timeout=2.0)

API reference

> =============

Reviewer: API is already generated elsewhere, you can link to it instead of re-generating here.

(probably this will lead to a lot of sphinx warnings about duplicate references?)

Reviewer: Examples: https://github.com/multiscale/muscle3/blob/develop/docs/source/tutorial.rst?plain=1#L93

.. _muscle-tester:

==========================
Testing your MUSCLE3 model
==========================

When developing a MUSCLE3 component, it is useful to be able to test it in
isolation — without having to set up and run the full coupled simulation. The
:class:`~libmuscle.pytest.MuscleTester` class (together with the
``muscle3_tester`` pytest fixture) makes this easy: it starts a real MUSCLE3
manager in the background and connects a *tester component* to all ports of
your implementation, so you can send and receive messages programmatically from

> within a pytest test.

Reviewer 2: Maybe point out that this allows automatic testing, as opposed to doing it by hand. To me, that's the main advantage, that you can essentially script the whole testing procedure and easily run it whenever you change something.

two forms.

.. tabs::

    .. group-tab:: Using a Path

        Place the yMMSL file in a ``tests/`` folder at the root of your project, next
        to your test files:

        .. code-block:: text

            my_project/
            ├── micro.py
            ├── micro.ymmsl
            └── tests/
                └── test_micro.py

        For example, for a simple micro model that receives a value on ``init`` and
        sends a result on ``final``:

        .. code-block:: yaml
            :caption: tests/micro.ymmsl

            ymmsl_version: v0.2

            programs:
              micro:
                ports:
                  f_init: init
                  o_f: final
                executable: python3
                args: /path/to/my_project/micro.py

            resources:
              micro:
                threads: 1

    .. group-tab:: Using an inline string

        You can embed the yMMSL configuration directly in your test file as a
        string.
        
        Tip: using ``Path(__file__).parent.resolve()`` gives the absolute path.

        .. code-block:: python
            :caption: tests/test_micro.py

            from pathlib import Path

>             PROJECT_DIR = Path(__file__).parent.parent.resolve()

Reviewer 2: or `.parents[1]` to avoid the repetition.

            PROJECT_DIR = Path(__file__).parent.parent.resolve()

            def test_micro_model(muscle3_tester: MuscleTester) -> None:
                """Test the micro model by acting as the macro."""
                tester = muscle3_tester.start_implementation(
                    PROJECT_DIR / "micro.ymmsl", "micro"
                    )

                # Send a message to the micro model's 'init' port
                tester.send("init", Message(0.0, 10.0, 42))

                # Receive the result from the micro model's 'final' port
                reply = tester.receive("final")

                assert reply.data == 42

    .. group-tab:: Using an inline string

        .. code-block:: python
            :caption: tests/test_micro.py

            from pathlib import Path
            from libmuscle import Message
            from libmuscle.pytest import MuscleTester

            TESTS_DIR = Path(__file__).parent.parent.resolve()

            CONFIG = f"""
            ymmsl_version: v0.2

            programs:
              micro:
                ports:
                  f_init: init
                  o_f: final
                executable: python3

>                 args: {PROJECT_DIR / "micro.py"}

Reviewer 2: Should that be TESTS_DIR instead of PROJECT_DIR, or vice versa?

Author: my_project/
            ├── micro.py
            ├── micro.ymmsl
            └── tests/
                └── test_micro.py

micro.py is located in PROJECT_DIR.

            resources:
              micro:
                threads: 1
            """

            def test_micro_model(muscle3_tester: MuscleTester) -> None:
                """Test the micro model by acting as the macro."""
                tester = muscle3_tester.start_implementation(CONFIG, "micro")

                # Send a message to the micro model's 'init' port
                tester.send("init", Message(0.0, 10.0, 42))

                # Receive the result from the micro model's 'final' port
                reply = tester.receive("final")

                assert reply.data == 42

Step 3: Run your tests
----------------------

Run ``pytest`` from the project root so that all paths resolve correctly:

.. code-block:: bash

    cd my_project
    pytest

> Timeouts and error handling

Reviewer 2: I like this, good thinking!

============================

By default, :meth:`~libmuscle.pytest.MuscleTester.start_implementation`
uses a 60-second timeout for all receive operations. If the implementation
does not send a message within that time, a :class:`RuntimeError` is raised
and the test fails.

You may want to increase the timeout when your implementation performs
expensive work between messages, for example, a micro-model that runs a
numerical solver for several minutes per time step. In those cases the default
60 seconds may expire before the component has had a chance to reply, causing a
spurious test failure.

Conversely, you may want to decrease the timeout in fast unit tests so
that a missing ``send`` call is detected quickly rather than making the test
suite hang for a full minute.

You can adjust this timeout:

.. code-block:: python

    tester = muscle3_tester.start_implementation(

>         TESTS_DIR / "micro.ymmsl", "micro", default_timeout=5.0

Reviewer 2: Also here TESTS vs PROJECT

Author: This should be indeed PROJECT

.. _muscle-tester:

==========================
Testing your MUSCLE3 model
==========================

When developing a MUSCLE3 component, it is useful to be able to test it in
isolation — without having to set up and run the full coupled simulation. The
:class:`~libmuscle.pytest.MuscleTester` class (together with the
``muscle3_tester`` pytest fixture) makes this easy: it starts a real MUSCLE3
manager in the background and connects a *tester component* to all ports of
your implementation, so you can send and receive messages programmatically from
within a pytest test.

.. note::

   The testing infrastructure described here relies on `pytest

>    <https://docs.pytest.org>`_, which is **not** installed automatically as

Reviewer: ```suggestion
   <https://docs.pytest.org>`_, which is not installed automatically as
```

   the same name to your test function. 

Quick start
===========

Step 1: Provide the yMMSL configuration for your implementation
---------------------------------------------------------------

:meth:`~libmuscle.pytest.MuscleTester.start_implementation` accepts the yMMSL in
two forms.

**Option A: yMMSL file**

Place the yMMSL file in a ``tests/`` folder at the root of your project, next
to your test files:

.. code-block:: text

    my_project/
    ├── micro.py
    └── tests/
        ├── micro.ymmsl
        └── test_micro.py

For example, for a simple micro model that receives a value on ``init`` and
sends a result on ``final``:

.. code-block:: yaml
    :caption: tests/micro.ymmsl

    ymmsl_version: v0.2

    programs:
      micro:
        ports:
          f_init: init
          o_f: final
        executable: python3
        args: micro.py

    resources:
      micro:
        threads: 1

.. note::
    The path passed to
    :meth:`~libmuscle.pytest.MuscleTester.start_implementation` is resolved
    relative to the directory from which you run ``pytest`` — typically the
    project root. The executable in the yMMSL file is also launched from that

>     directory, so make sure your implementation is importable from there.

Reviewer: Did you test this? I'm not sure if this is correct: in a regular muscle3 simulation the python executable is started inside the run directory.

For these tests that run directory would be somewhere in a temporary folder, and `micro.py` may not be found like this.

Author: You are right, is this solved if you give an absolute path for the script in args?

**Option B: inline yMMSL string**
You can embed the yMMSL configuration directly in your test file as a string.
This is handy for short configurations or when you want the test to be fully
self-contained:

.. code-block:: python
    :caption: tests/test_micro.py

    CONFIG = """
    ymmsl_version: v0.2

    programs:
      micro:
        ports:
          f_init: init
          o_f: final
        executable: python3
        args: micro.py

    resources:
      micro:
        threads: 1
    """

Step 2: Use the ``muscle3_tester`` fixture in your test
-------------------------------------------------------

> **Using a Path:**

Reviewer: Might be nice to use tabs for the two cases? See the dynamic port configuration PR for some examples: https://github.com/multiscale/muscle3/pull/350/changes#diff-509659d28e6e781efc52de54ecdf261fd94227761dbc10004b7d3d24a95a4460

    CONFIG = """
    ymmsl_version: v0.2

    programs:
      micro:
        ports:
          f_init: init
          o_f: final
        executable: python3
        args: micro.py

    resources:
      micro:
        threads: 1
    """

    def test_micro_model(muscle3_tester: MuscleTester) -> None:
        """Test the micro model by acting as the macro."""
        tester = muscle3_tester.start_implementation(CONFIG, "micro")

        # Send a message to the micro model's 'init' port
        tester.send("init", Message(0.0, 10.0, 42))

        # Receive the result from the micro model's 'final' port
        reply = tester.receive("final")

        assert reply.data == 42

Step 3: Run your tests
----------------------

Run ``pytest`` from the project root so that all paths resolve correctly:

.. code-block:: bash

    cd my_project
    pytest

Timeouts and error handling
============================

By default, :meth:`~libmuscle.pytest.MuscleTester.start_implementation`
uses a 60-second timeout for all receive operations. If the implementation
does not send a message within that time, a :class:`RuntimeError` is raised
and the test fails.

> You may want to **increase** the timeout when your implementation performs

Reviewer: No need to emphasize
```suggestion
You may want to increase the timeout when your implementation performs
```

expensive work between messages, for example, a micro-model that runs a
numerical solver for several minutes per time step. In those cases the default
60 seconds may expire before the component has had a chance to reply, causing a
spurious test failure.

> Conversely, you may want to **decrease** the timeout in fast unit tests so

Reviewer: ```suggestion
Conversely, you may want to decrease the timeout in fast unit tests so
```

that a missing ``send`` call is detected quickly rather than making the test
suite hang for a full minute.

You can adjust this timeout:

.. code-block:: python

    tester = muscle3_tester.start_implementation(
        TESTS_DIR / "micro.ymmsl", "micro", default_timeout=5.0
    )

You can also override the timeout for individual receive calls:

.. code-block:: python

    reply = tester.receive("final", timeout=2.0)

API reference
=============

The full API for :class:`libmuscle.pytest.MuscleTester` and

> :class:`libmuscle.pytest.implementation_tester.ImplementationTester` is

Reviewer: You could prefix this with `~` so it renders just the final part of the dotted name in the HTML docs.

```suggestion
The full API for :class:`~libmuscle.pytest.MuscleTester` and
:class:`~libmuscle.pytest.implementation_tester.ImplementationTester` is
```

===========

Step 1: Provide the yMMSL configuration for your implementation
---------------------------------------------------------------

:meth:`~libmuscle.pytest.MuscleTester.start_implementation` accepts the yMMSL in
two forms.

.. tabs::

    .. group-tab:: Using a Path

        Place the yMMSL file in a ``tests/`` folder at the root of your project, next
        to your test files:

        .. code-block:: text

            my_project/
            ├── micro.py
            ├── micro.ymmsl
            └── tests/
                └── test_micro.py

        For example, for a simple micro model that receives a value on ``init`` and
        sends a result on ``final``:

        .. code-block:: yaml
            :caption: tests/micro.ymmsl

            ymmsl_version: v0.2

            programs:
              micro:
                ports:
                  f_init: init
                  o_f: final
                executable: python3
                args: /path/to/my_project/micro.py

            resources:
              micro:
                threads: 1

    .. group-tab:: Using an inline string

        You can embed the yMMSL configuration directly in your test file as an

>         f-string. Using ``Path(__file__).parent.resolve()`` gives the absolute path

Reviewer: It doesn't have to be an f-string, any string would do.

```suggestion
        You can embed the yMMSL configuration directly in your test file as a
        string.
        
        Tip: using ``Path(__file__).parent.resolve()`` gives the absolute path
```
